### ISSUE TITLE
One Tip For Project

### ISSUE THREAD

**Issue by ff1Zzd** [nus-cs2030/2122-s1#242]
For those who have not finished the project, please do not schedule ServeEvent from WaitEvent!!! You may find a way to pass level 3 if you are implementing this way, however you will suffer a lot when you reach level5. 

As what has been suggested by others, only initiate a serveEvent when you need to.

**Comment by brandonrhan** [comment]
I totally agree!!! A way to do this is to create an event between wait and serve, and ask it to handle the check. Another tip will be update the next available time only in Serve event. Otherwise you need to change the structure a lot when you reach level 5.

**Comment by PokezardVGC** [comment]
Hello, one of the jokers here trying to last min start on the project. To charity, we don't generate a wait event, just create a serveEvent, incremented by the wait time?

**Comment by brandonrhan** [comment]
> Hello, one of the jokers here trying to last min start on the project. To charity, we don't generate a wait event, just create a serveEvent, incremented by the wait time?

It might work but I suggest that try to follow the flow of Arrive -> Wait -> Serve -> Done, and increment the wait time only at the Serve Event. It will be easier for you to do the checks, especially when more than one customer is waiting

**Comment by weiquany** [comment]
I agree with what @brandonrhan mentioned. Ultimately, a WaitEvent will transition into a ServeEvent, hence I do not think that there is a problem scheduling ServeEvent from WaitEvent as long as it is only initiated when the customer is going to be served.

**Comment by mewrion** [comment]
Another clown here! For level 5, is it right to overload constructor for the same simulator? Judging from the level itself, seems like an extra simulator class would be better, rather than editing level 1 to 4

**Comment by brandonrhan** [comment]
> Another clown here! For level 5, is it right to overload constructor for the same simulator? Judging from the level itself, seems like an extra simulator class would be better, rather than editing level 1 to 4

actually, if you want you can create a different simualtor12345 correspond to Main12345 for each level

**Comment by ff1Zzd** [comment]
> I agree with what @brandonrhan mentioned. Ultimately, a WaitEvent will transition into a ServeEvent, hence I do not think that there is a problem scheduling ServeEvent from WaitEvent as long as it is only initiated when the customer is going to be served.

Perhaps I didnt make myself clear enough. I guess it may work depends on how you implement the code. In my original code my WaitEvent will execute at the time that the customer arrives(when they arrive they find out that they need wait), and then schedule a ServeEvent for the customer based on customer's serviceTime, so its like "schedule" a future ServeEvent in a sense.

In this case, I believe it will not pass level 5 if we implement it that way. Anyway, as what you said, no matter how we implement our rest of code, the important thing is that we only create the ServeEvent when the customer is going to be served.

**Comment by ff1Zzd** [comment]
> Hello, one of the jokers here trying to last min start on the project. To charity, we don't generate a wait event, just create a serveEvent, incremented by the wait time?

I would still suggest to create a WaitEvent so it will be easier for you to track your logic flow when you debug, it makes you flow clearer. The key here is that we should only create the ServeEvent when there is a customer to be served instead of creating it in advance based on customer's service time (like what I did, and believe me its really painful to change that implementation if you wanna pass level 5 so adjust it early). The hints i can give is that think of the moments that customers will be served immediately. (i.e. When customer arrives and find a idle server or when a server finish his resting and there are customers waiting in his queue,,,)

**Comment by dragane359** [comment]
For level 5, since we cannot tie the service time to the customer, would it be okay to treat it in a similar way to `restTimes` (i.e. generate all `n` values of service time at the start of the simulation where `n` is the number of customers and get the customer `serviceTime` by looking at the index of the number of customers served in the `serviceTimes` list?)

**Comment by ff1Zzd** [comment]
> For level 5, since we cannot tie the service time to the customer, would it be okay to treat it in a similar way to `restTimes` (i.e. generate all `n` values of service time at the start of the simulation where `n` is the number of customers and get the customer `serviceTime` by looking at the index of the number of customers served in the `serviceTimes` list?)

I dun think that works, nth customer being served may not be the nth customer arrive. So you can’t generate the service time at the start

**Comment by dragane359** [comment]
> > For level 5, since we cannot tie the service time to the customer, would it be okay to treat it in a similar way to `restTimes` (i.e. generate all `n` values of service time at the start of the simulation where `n` is the number of customers and get the customer `serviceTime` by looking at the index of the number of customers served in the `serviceTimes` list?)
> 
> I dun think that works, nth customer being served may not be the nth customer arrive. So you can’t generate the service time at the start

hmm, how about if we generate all service times at the start, and keep a count tracker for the number of customers served. then could we use the value of the count we tracked to get the index of `serviceTime` in the serviceTimes array, so the nth customer would not neccesarily get the nth `serviceTime`, but based on the order it's served?

**Comment by brandonrhan** [comment]
> > > For level 5, since we cannot tie the service time to the customer, would it be okay to treat it in a similar way to `restTimes` (i.e. generate all `n` values of service time at the start of the simulation where `n` is the number of customers and get the customer `serviceTime` by looking at the index of the number of customers served in the `serviceTimes` list?)
> > 
> > 
> > I dun think that works, nth customer being served may not be the nth customer arrive. So you can’t generate the service time at the start
> 
> hmm, how about if we generate all service times at the start, and keep a count tracker for the number of customers served. then could we use the value of the count we tracked to get the index of `serviceTime` in the serviceTimes array, so the nth customer would not neccesarily get the nth `serviceTime`, but based on the order it's served?

i think it would work if you use an static attribute to count every serve and only generate the serviceTime in Serve Event.

**Comment by cadencjk** [comment]
> > > For level 5, since we cannot tie the service time to the customer, would it be okay to treat it in a similar way to `restTimes` (i.e. generate all `n` values of service time at the start of the simulation where `n` is the number of customers and get the customer `serviceTime` by looking at the index of the number of customers served in the `serviceTimes` list?)
> > 
> > 
> > I dun think that works, nth customer being served may not be the nth customer arrive. So you can’t generate the service time at the start
> 
> hmm, how about if we generate all service times at the start, and keep a count tracker for the number of customers served. then could we use the value of the count we tracked to get the index of `serviceTime` in the serviceTimes array, so the nth customer would not neccesarily get the nth `serviceTime`, but based on the order it's served?

What I did was I passed the random generator down the events, to the serve event. And based on the numbers of customers served, the event then triggers a Done Serving event with the randomly generated value.

**Comment by ff1Zzd** [comment]
> > > For level 5, since we cannot tie the service time to the customer, would it be okay to treat it in a similar way to `restTimes` (i.e. generate all `n` values of service time at the start of the simulation where `n` is the number of customers and get the customer `serviceTime` by looking at the index of the number of customers served in the `serviceTimes` list?)
> > 
> > 
> > I dun think that works, nth customer being served may not be the nth customer arrive. So you can’t generate the service time at the start
> 
> hmm, how about if we generate all service times at the start, and keep a count tracker for the number of customers served. then could we use the value of the count we tracked to get the index of `serviceTime` in the serviceTimes array, so the nth customer would not neccesarily get the nth `serviceTime`, but based on the order it's served?

I think this is possible, coz basically you only know the serviceTime when your customer gets served right.

**Comment by huzaifa1712** [comment]
> For level 5, since we cannot tie the service time to the customer, would it be okay to treat it in a similar way to `restTimes` (i.e. generate all `n` values of service time at the start of the simulation where `n` is the number of customers and get the customer `serviceTime` by looking at the index of the number of customers served in the `serviceTimes` list?)

I think you can, but you might find it easier to use suppliers and caching for this. What I did was have a SvcTimeCache class that basically behaves like Lazy.java with a get method and supplier (of double svc time), but once the supplier is called it doesn't get again. Then the Customer has a SvcTimeCache and when creating customers in the constructor I pass in the suppliers with () -> rng.genServiceTime() for level 5. For previous levels I create the suppliers in the Main method and pass in with another overloaded constructor.

**Comment by kxshxsh** [comment]
> > Another clown here! For level 5, is it right to overload constructor for the same simulator? Judging from the level itself, seems like an extra simulator class would be better, rather than editing level 1 to 4
> 
> actually, if you want you can create a different simualtor12345 correspond to Main12345 for each level

Best advice^ I just have simulatorN.java and MainN.java per level!
